{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.13.2'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os \n",
    "import sys\n",
    "from subprocess import run\n",
    "import importlib\n",
    "import mixing\n",
    "importlib.reload(mixing)\n",
    "import glob\n",
    "from adjustText import adjust_text\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "plt.rcParams[\"figure.figsize\"] = (4,4)\n",
    "sns.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1) Run this to construct long read assemblies\n",
    "RUN_FLY_BASH = '''\n",
    "#!/bin/bash \n",
    "#SBATCH --job-name=checkM\n",
    "#SBATCH --time=4:00:00\n",
    "#SBATCH --mem=96G\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --exclude=m005\n",
    "#SBATCH --account pmg\n",
    "#SBATCH --output=/burg/pmg/users/ic2465/Projects/MANU_copangraph/KmerMixing/flye-%x-%j.log\n",
    "\n",
    "flye --meta --threads 16 --pacbio-hifi ${SAMPLE} -o ${OUTDIR}\n",
    "'''\n",
    "\n",
    "# 2) Select all contigs in all long read assemblies with length > 1M\n",
    "# python split_and_filter_long_reads.py <ASSEMBLY> 1 <CANDIATE_GENOMES_OUTDIR>\n",
    "\n",
    "# 3) Run checkM on all contigs > 1M \n",
    "RUN_CHECKM_BASH = '''\n",
    "#!/bin/bash \n",
    "#SBATCH --job-name=checkM\n",
    "#SBATCH --time=4:00:00\n",
    "#SBATCH --mem=90G\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --exclude=m005\n",
    "#SBATCH --account pmg\n",
    "#SBATCH --output=/burg/pmg/users/ic2465/Projects/MANU_copangraph/KmerMixing/checkM-%x-%j.log\n",
    "checkm lineage_wf -x fa $DIR ${DIR}/checkm_out -t 16 --file output.txt\n",
    "'''\n",
    "\n",
    "# 4) Select contigs with Contamination < 5%\n",
    "# checkm qa checkm_out/lineage.ms checkm_out -o 2 --tab_table --file checkm_output.tsv\n",
    "# then filter table to contamination < 5 in ipython. Saved to Bin_contamination_lt5.csv\n",
    "\n",
    "# 5) Run dRep at 99, 97, 95 % ANI\n",
    "# /usr/bin/time -v dRep dereplicate -g low_contam_bins/*.fa -p 32 --ignoreGenomeQuality -sa 0.95 --clusterAlg average dreped_bins_95\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_536058/2446360529.py:8: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  cluster_reps = full_data.groupby('secondary_cluster').apply(lambda x: x.loc[x.completeness == x.completeness.max(),:].iloc[0,:]).reset_index(drop=True)\n"
     ]
    }
   ],
   "source": [
    "# 6) select representatives\n",
    "ANI=95\n",
    "drep_data = pd.read_csv(f'../data/KmerMixing/gherig/bins/dreped_bins_{ANI}/data_tables/Cdb.csv')\n",
    "checkm_data = pd.read_csv('../data/KmerMixing/gherig/bins/checkm_out/Bins_contamination_lt5.csv')\n",
    "drep_data['bin_id'] = drep_data.genome.apply(lambda x: os.path.splitext(x)[0])\n",
    "full_data = pd.merge(checkm_data, drep_data, on='bin_id')\n",
    "\n",
    "cluster_reps = full_data.groupby('secondary_cluster').apply(lambda x: x.loc[x.completeness == x.completeness.max(),:].iloc[0,:]).reset_index(drop=True)\n",
    "cluster_reps.to_csv(f'../data/KmerMixing/gherig/bins/drep_representatives_{ANI}/cluster_reps.csv',index=None)\n",
    "# 7) move genomes to representative dir\n",
    "ALL_BINS= '/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/KmerMixing/gherig/bins/contigs_gt_1M'\n",
    "REP_BINS= f'/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/KmerMixing/gherig/bins/drep_representatives_{ANI}'\n",
    "for b in cluster_reps.bin_id:\n",
    "    cmd = 'cp ' + os.path.join(ALL_BINS, f'{b}.fa') + ' ' + os.path.join(REP_BINS, f'{b}.fa')\n",
    "    run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7) call gherig run to launch the seq in genome analysii\n",
    "RUN_GHERIG_MIXING='''#!/bin/bash \n",
    "#SBATCH --job-name=kmer-mixing\n",
    "#SBATCH --time=48:00:00\n",
    "#SBATCH --mem=128G\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --exclude=m005\n",
    "#SBATCH --account pmg\n",
    "#SBATCH --output=/burg/pmg/users/ic2465/Projects/MANU_copangraph/KmerMixing/repmix-%x-%j.log\n",
    "python replicate_run.py ${r} ${g}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# - some useful paths\n",
    "ANI = 99\n",
    "#PREF=f'hmlgy_panel_gherig_mag{ANI}_'\n",
    "PREF_COV=f'hmlgy_panel_coverage_magani{ANI}_'\n",
    "PREF_MULTI=f'hmlgy_panel_multigenome_magani{ANI}_'\n",
    "OUTDIR='/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/KmerMixing/gherig/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8) preprocess raw output\n",
    "get_base = lambda x: os.path.splitext(os.path.basename(x))[0]\n",
    "data = pd.read_csv(f'../data/KmerMixing/gherig/ani_{ANI}_mixing_results.csv', index_col=0)\n",
    "data['file'] = data.tool\n",
    "data['genome_ani'] = ANI\n",
    "data.tool = data.file.apply(lambda x: 'copangraph' if 'copangraph' in get_base(x) else 'megahit')\n",
    "data['parameter'] = data.file.apply(lambda x: float(get_base(x)[1:]) if 'copangraph' not in get_base(x) else float(get_base(x).split('_')[3]))\n",
    "data\n",
    "plot_dat = data.groupby(['parameter', 'tool', 'file', 'genome_ani']).apply(\n",
    "    lambda x: pd.Series(\n",
    "        [\n",
    "            x.nid.shape[0],  # how many nodes?\n",
    "            x.total_bp.sum(), # total bp in graph?\n",
    "            x.max_bp.sum(),  # total bp in graph, if we only take the longest seq in each node?\n",
    "            x.is_multi_genome.sum()/x.shape[0],  # proportion of multi genome nodes?\n",
    "            x.is_single_genome.sum()/x.shape[0],  # proportion of single genome nodes?\n",
    "            (x.unmapped_bp == x.total_bp).sum()/x.shape[0], # proportion of unmapped nodes\n",
    "            x.num_genomes_in.sum()/x.shape[0],  # average number of genomes per node?\n",
    "            x.has_unmapped.sum()/x.shape[0], # proportion of nodes containing some unmapped sequence?\n",
    "            x.unmapped_bp.sum(), # sum total unmapped\n",
    "            x.unmapped_bp.sum()/x.total_bp.sum()# propotion bp unmappable\n",
    "        ],\n",
    "        index=[\n",
    "            'num_nodes', 'total_bp_in_graph', 'sum_max_seq_per_node', \n",
    "            'multi_genome_node_proportion', 'single_genome_node_proportion', 'unmapped_node_proportion',\n",
    "            'mean_num_genomes_per_node', 'node_has_unmapped_sequences_proportion', \n",
    "            'total_unmapped_bp', 'proportion_bp_unmappable'\n",
    "        ]\n",
    "    )\n",
    ").reset_index()\n",
    "plot_dat['proportion_bp_mappable'] = 1-plot_dat.proportion_bp_unmappable\n",
    "print('n50 calc...')\n",
    "plot_dat['n50'] = plot_dat.file.apply(lambda x: mixing.n50(x))\n",
    "print('num edge calc...')\n",
    "plot_dat['num_edges'] = plot_dat.file.apply(lambda x: mixing.num_edges(x))\n",
    "plot_dat.to_csv(f'ani_{ANI}_plot_dat.csv', index=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parameter</th>\n",
       "      <th>tool</th>\n",
       "      <th>file</th>\n",
       "      <th>genome_ani</th>\n",
       "      <th>num_nodes</th>\n",
       "      <th>total_bp_in_graph</th>\n",
       "      <th>sum_max_seq_per_node</th>\n",
       "      <th>multi_genome_node_proportion</th>\n",
       "      <th>single_genome_node_proportion</th>\n",
       "      <th>unmapped_node_proportion</th>\n",
       "      <th>mean_num_genomes_per_node</th>\n",
       "      <th>node_has_unmapped_sequences_proportion</th>\n",
       "      <th>total_unmapped_bp</th>\n",
       "      <th>proportion_bp_unmappable</th>\n",
       "      <th>proportion_bp_mappable</th>\n",
       "      <th>n50</th>\n",
       "      <th>num_edges</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.005</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1203098.0</td>\n",
       "      <td>1.340150e+09</td>\n",
       "      <td>1.058038e+09</td>\n",
       "      <td>0.044872</td>\n",
       "      <td>0.182432</td>\n",
       "      <td>0.772696</td>\n",
       "      <td>0.325932</td>\n",
       "      <td>0.790403</td>\n",
       "      <td>9.996008e+08</td>\n",
       "      <td>0.745887</td>\n",
       "      <td>0.254113</td>\n",
       "      <td>2279</td>\n",
       "      <td>275751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.010</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1223090.0</td>\n",
       "      <td>1.339838e+09</td>\n",
       "      <td>9.556943e+08</td>\n",
       "      <td>0.055460</td>\n",
       "      <td>0.187229</td>\n",
       "      <td>0.757311</td>\n",
       "      <td>0.357938</td>\n",
       "      <td>0.791212</td>\n",
       "      <td>9.682273e+08</td>\n",
       "      <td>0.722645</td>\n",
       "      <td>0.277355</td>\n",
       "      <td>1771</td>\n",
       "      <td>316363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.020</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1243165.0</td>\n",
       "      <td>1.339523e+09</td>\n",
       "      <td>8.688648e+08</td>\n",
       "      <td>0.069864</td>\n",
       "      <td>0.190843</td>\n",
       "      <td>0.739294</td>\n",
       "      <td>0.399913</td>\n",
       "      <td>0.792199</td>\n",
       "      <td>9.422581e+08</td>\n",
       "      <td>0.703428</td>\n",
       "      <td>0.296572</td>\n",
       "      <td>1472</td>\n",
       "      <td>355628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.030</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1250211.0</td>\n",
       "      <td>1.339418e+09</td>\n",
       "      <td>8.480685e+08</td>\n",
       "      <td>0.075128</td>\n",
       "      <td>0.191916</td>\n",
       "      <td>0.732955</td>\n",
       "      <td>0.416946</td>\n",
       "      <td>0.792299</td>\n",
       "      <td>9.352227e+08</td>\n",
       "      <td>0.698231</td>\n",
       "      <td>0.301769</td>\n",
       "      <td>1398</td>\n",
       "      <td>369122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.040</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1253397.0</td>\n",
       "      <td>1.339377e+09</td>\n",
       "      <td>8.396531e+08</td>\n",
       "      <td>0.077400</td>\n",
       "      <td>0.192396</td>\n",
       "      <td>0.730204</td>\n",
       "      <td>0.426151</td>\n",
       "      <td>0.792300</td>\n",
       "      <td>9.316509e+08</td>\n",
       "      <td>0.695585</td>\n",
       "      <td>0.304415</td>\n",
       "      <td>1369</td>\n",
       "      <td>375111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.050</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1255173.0</td>\n",
       "      <td>1.339359e+09</td>\n",
       "      <td>8.345463e+08</td>\n",
       "      <td>0.078926</td>\n",
       "      <td>0.192495</td>\n",
       "      <td>0.728579</td>\n",
       "      <td>0.433617</td>\n",
       "      <td>0.792393</td>\n",
       "      <td>9.293850e+08</td>\n",
       "      <td>0.693903</td>\n",
       "      <td>0.306097</td>\n",
       "      <td>1350</td>\n",
       "      <td>378454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.060</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1256473.0</td>\n",
       "      <td>1.339349e+09</td>\n",
       "      <td>8.314314e+08</td>\n",
       "      <td>0.079968</td>\n",
       "      <td>0.192484</td>\n",
       "      <td>0.727548</td>\n",
       "      <td>0.438847</td>\n",
       "      <td>0.792535</td>\n",
       "      <td>9.277911e+08</td>\n",
       "      <td>0.692718</td>\n",
       "      <td>0.307282</td>\n",
       "      <td>1336</td>\n",
       "      <td>380860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.100</td>\n",
       "      <td>copangraph</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1259308.0</td>\n",
       "      <td>1.339334e+09</td>\n",
       "      <td>8.259228e+08</td>\n",
       "      <td>0.082150</td>\n",
       "      <td>0.192347</td>\n",
       "      <td>0.725503</td>\n",
       "      <td>0.450269</td>\n",
       "      <td>0.792704</td>\n",
       "      <td>9.229110e+08</td>\n",
       "      <td>0.689082</td>\n",
       "      <td>0.310918</td>\n",
       "      <td>1306</td>\n",
       "      <td>385786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>21.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>21383619.0</td>\n",
       "      <td>1.432286e+09</td>\n",
       "      <td>1.432286e+09</td>\n",
       "      <td>0.110418</td>\n",
       "      <td>0.202550</td>\n",
       "      <td>0.687032</td>\n",
       "      <td>0.535583</td>\n",
       "      <td>0.687032</td>\n",
       "      <td>1.021420e+09</td>\n",
       "      <td>0.713140</td>\n",
       "      <td>0.286860</td>\n",
       "      <td>103</td>\n",
       "      <td>24414695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>29.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>5136069.0</td>\n",
       "      <td>1.045655e+09</td>\n",
       "      <td>1.045655e+09</td>\n",
       "      <td>0.054212</td>\n",
       "      <td>0.175409</td>\n",
       "      <td>0.770379</td>\n",
       "      <td>0.317683</td>\n",
       "      <td>0.770379</td>\n",
       "      <td>8.241828e+08</td>\n",
       "      <td>0.788198</td>\n",
       "      <td>0.211802</td>\n",
       "      <td>295</td>\n",
       "      <td>3235339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>39.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>2885331.0</td>\n",
       "      <td>9.578978e+08</td>\n",
       "      <td>9.578978e+08</td>\n",
       "      <td>0.030338</td>\n",
       "      <td>0.129319</td>\n",
       "      <td>0.840343</td>\n",
       "      <td>0.210971</td>\n",
       "      <td>0.840343</td>\n",
       "      <td>8.030602e+08</td>\n",
       "      <td>0.838357</td>\n",
       "      <td>0.161643</td>\n",
       "      <td>571</td>\n",
       "      <td>1317886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>59.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1583700.0</td>\n",
       "      <td>8.641560e+08</td>\n",
       "      <td>8.641560e+08</td>\n",
       "      <td>0.034486</td>\n",
       "      <td>0.138475</td>\n",
       "      <td>0.827039</td>\n",
       "      <td>0.241070</td>\n",
       "      <td>0.827039</td>\n",
       "      <td>7.459140e+08</td>\n",
       "      <td>0.863171</td>\n",
       "      <td>0.136829</td>\n",
       "      <td>1122</td>\n",
       "      <td>723091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>79.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>1207241.0</td>\n",
       "      <td>8.370560e+08</td>\n",
       "      <td>8.370560e+08</td>\n",
       "      <td>0.043304</td>\n",
       "      <td>0.145823</td>\n",
       "      <td>0.810874</td>\n",
       "      <td>0.278799</td>\n",
       "      <td>0.810874</td>\n",
       "      <td>7.348289e+08</td>\n",
       "      <td>0.877873</td>\n",
       "      <td>0.122127</td>\n",
       "      <td>1621</td>\n",
       "      <td>605820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>99.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>945004.0</td>\n",
       "      <td>8.113629e+08</td>\n",
       "      <td>8.113629e+08</td>\n",
       "      <td>0.046078</td>\n",
       "      <td>0.150886</td>\n",
       "      <td>0.803036</td>\n",
       "      <td>0.290744</td>\n",
       "      <td>0.803036</td>\n",
       "      <td>7.210545e+08</td>\n",
       "      <td>0.888695</td>\n",
       "      <td>0.111305</td>\n",
       "      <td>2150</td>\n",
       "      <td>482043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>119.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>893264.0</td>\n",
       "      <td>8.271725e+08</td>\n",
       "      <td>8.271725e+08</td>\n",
       "      <td>0.043996</td>\n",
       "      <td>0.147832</td>\n",
       "      <td>0.808172</td>\n",
       "      <td>0.278551</td>\n",
       "      <td>0.808172</td>\n",
       "      <td>7.403534e+08</td>\n",
       "      <td>0.895041</td>\n",
       "      <td>0.104959</td>\n",
       "      <td>2329</td>\n",
       "      <td>422941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>141.000</td>\n",
       "      <td>megahit</td>\n",
       "      <td>/burg/pmg/users/ic2465/Projects/MANU_copangrap...</td>\n",
       "      <td>99</td>\n",
       "      <td>897378.0</td>\n",
       "      <td>8.554268e+08</td>\n",
       "      <td>8.554268e+08</td>\n",
       "      <td>0.042893</td>\n",
       "      <td>0.148003</td>\n",
       "      <td>0.809104</td>\n",
       "      <td>0.271676</td>\n",
       "      <td>0.809104</td>\n",
       "      <td>7.694047e+08</td>\n",
       "      <td>0.899440</td>\n",
       "      <td>0.100560</td>\n",
       "      <td>2388</td>\n",
       "      <td>389960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    parameter        tool                                               file  \\\n",
       "0       0.005  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "1       0.010  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "2       0.020  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "3       0.030  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "4       0.040  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "5       0.050  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "6       0.060  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "7       0.100  copangraph  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "8      21.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "9      29.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "10     39.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "11     59.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "12     79.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "13     99.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "14    119.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "15    141.000     megahit  /burg/pmg/users/ic2465/Projects/MANU_copangrap...   \n",
       "\n",
       "    genome_ani   num_nodes  total_bp_in_graph  sum_max_seq_per_node  \\\n",
       "0           99   1203098.0       1.340150e+09          1.058038e+09   \n",
       "1           99   1223090.0       1.339838e+09          9.556943e+08   \n",
       "2           99   1243165.0       1.339523e+09          8.688648e+08   \n",
       "3           99   1250211.0       1.339418e+09          8.480685e+08   \n",
       "4           99   1253397.0       1.339377e+09          8.396531e+08   \n",
       "5           99   1255173.0       1.339359e+09          8.345463e+08   \n",
       "6           99   1256473.0       1.339349e+09          8.314314e+08   \n",
       "7           99   1259308.0       1.339334e+09          8.259228e+08   \n",
       "8           99  21383619.0       1.432286e+09          1.432286e+09   \n",
       "9           99   5136069.0       1.045655e+09          1.045655e+09   \n",
       "10          99   2885331.0       9.578978e+08          9.578978e+08   \n",
       "11          99   1583700.0       8.641560e+08          8.641560e+08   \n",
       "12          99   1207241.0       8.370560e+08          8.370560e+08   \n",
       "13          99    945004.0       8.113629e+08          8.113629e+08   \n",
       "14          99    893264.0       8.271725e+08          8.271725e+08   \n",
       "15          99    897378.0       8.554268e+08          8.554268e+08   \n",
       "\n",
       "    multi_genome_node_proportion  single_genome_node_proportion  \\\n",
       "0                       0.044872                       0.182432   \n",
       "1                       0.055460                       0.187229   \n",
       "2                       0.069864                       0.190843   \n",
       "3                       0.075128                       0.191916   \n",
       "4                       0.077400                       0.192396   \n",
       "5                       0.078926                       0.192495   \n",
       "6                       0.079968                       0.192484   \n",
       "7                       0.082150                       0.192347   \n",
       "8                       0.110418                       0.202550   \n",
       "9                       0.054212                       0.175409   \n",
       "10                      0.030338                       0.129319   \n",
       "11                      0.034486                       0.138475   \n",
       "12                      0.043304                       0.145823   \n",
       "13                      0.046078                       0.150886   \n",
       "14                      0.043996                       0.147832   \n",
       "15                      0.042893                       0.148003   \n",
       "\n",
       "    unmapped_node_proportion  mean_num_genomes_per_node  \\\n",
       "0                   0.772696                   0.325932   \n",
       "1                   0.757311                   0.357938   \n",
       "2                   0.739294                   0.399913   \n",
       "3                   0.732955                   0.416946   \n",
       "4                   0.730204                   0.426151   \n",
       "5                   0.728579                   0.433617   \n",
       "6                   0.727548                   0.438847   \n",
       "7                   0.725503                   0.450269   \n",
       "8                   0.687032                   0.535583   \n",
       "9                   0.770379                   0.317683   \n",
       "10                  0.840343                   0.210971   \n",
       "11                  0.827039                   0.241070   \n",
       "12                  0.810874                   0.278799   \n",
       "13                  0.803036                   0.290744   \n",
       "14                  0.808172                   0.278551   \n",
       "15                  0.809104                   0.271676   \n",
       "\n",
       "    node_has_unmapped_sequences_proportion  total_unmapped_bp  \\\n",
       "0                                 0.790403       9.996008e+08   \n",
       "1                                 0.791212       9.682273e+08   \n",
       "2                                 0.792199       9.422581e+08   \n",
       "3                                 0.792299       9.352227e+08   \n",
       "4                                 0.792300       9.316509e+08   \n",
       "5                                 0.792393       9.293850e+08   \n",
       "6                                 0.792535       9.277911e+08   \n",
       "7                                 0.792704       9.229110e+08   \n",
       "8                                 0.687032       1.021420e+09   \n",
       "9                                 0.770379       8.241828e+08   \n",
       "10                                0.840343       8.030602e+08   \n",
       "11                                0.827039       7.459140e+08   \n",
       "12                                0.810874       7.348289e+08   \n",
       "13                                0.803036       7.210545e+08   \n",
       "14                                0.808172       7.403534e+08   \n",
       "15                                0.809104       7.694047e+08   \n",
       "\n",
       "    proportion_bp_unmappable  proportion_bp_mappable   n50  num_edges  \n",
       "0                   0.745887                0.254113  2279     275751  \n",
       "1                   0.722645                0.277355  1771     316363  \n",
       "2                   0.703428                0.296572  1472     355628  \n",
       "3                   0.698231                0.301769  1398     369122  \n",
       "4                   0.695585                0.304415  1369     375111  \n",
       "5                   0.693903                0.306097  1350     378454  \n",
       "6                   0.692718                0.307282  1336     380860  \n",
       "7                   0.689082                0.310918  1306     385786  \n",
       "8                   0.713140                0.286860   103   24414695  \n",
       "9                   0.788198                0.211802   295    3235339  \n",
       "10                  0.838357                0.161643   571    1317886  \n",
       "11                  0.863171                0.136829  1122     723091  \n",
       "12                  0.877873                0.122127  1621     605820  \n",
       "13                  0.888695                0.111305  2150     482043  \n",
       "14                  0.895041                0.104959  2329     422941  \n",
       "15                  0.899440                0.100560  2388     389960  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 9) plot data\n",
    "plot_dat = pd.read_csv('ani_99_plot_dat.csv')\n",
    "plot_dat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_unlabelled_version(ax, name, tight_layout=True):\n",
    "    name = os.path.splitext(name)[0]\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    frame1 = plt.gca()\n",
    "    frame1.legend().set_visible(False)\n",
    "    if tight_layout:\n",
    "        plt.tight_layout()\n",
    "    plt.savefig(f'{name}_UNLBLD.pdf', dpi=1400, bbox_inches='tight')\n",
    "    plt.savefig(f'{name}_UNLBLD.png', dpi=900, bbox_inches='tight')\n",
    "    plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# SUPPLEMENTARY FIGURES SHOWING PROPERTIES BY PARAMETERS\n",
    "\n",
    "for prop, lims in [\n",
    "    ('mean_num_genomes_per_node', (0.15, 0.6)), \n",
    "    ('n50', (0, 3000)), \n",
    "    #('total_bp_in_graph', (0.75*1e9, 1.5*1e9)), \n",
    "    #('num_nodes', (5*1e5, 2.2*1e7)), \n",
    "    #('num_edges', (2*1e5, 25*1e6))\n",
    "    ('total_bp_in_graph', None), \n",
    "    ('num_nodes', None), \n",
    "    ('num_edges', None)\n",
    "]:\n",
    "\n",
    "    # megahit mean genomes per node\n",
    "    ax=sns.lineplot(x=plot_dat.loc[plot_dat.tool == 'megahit', :].parameter, y=plot_dat.loc[plot_dat.tool == 'megahit', :][prop], color='red')\n",
    "    if lims is not None:\n",
    "        ax.set_ylim(*lims)\n",
    "    ax.set_xticks(plot_dat.loc[plot_dat.tool == 'megahit'].parameter.sort_values())\n",
    "    name = OUTDIR + PREF + f'megahit_{prop}.pdf'\n",
    "    plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "    plot_unlabelled_version(ax, name)\n",
    "\n",
    "    # copangraph mean genomes per node\n",
    "    ax=sns.lineplot(x=plot_dat.loc[plot_dat.tool == 'copangraph', :].parameter, y=plot_dat.loc[plot_dat.tool == 'copangraph', :][prop], color='blue')\n",
    "    if lims is not None:\n",
    "        ax.set_ylim(*lims)\n",
    "    ax.set_xticks(plot_dat.loc[plot_dat.tool == 'copangraph'].parameter.sort_values())\n",
    "    name = OUTDIR + PREF + f'copangraph_{prop}.pdf'\n",
    "    plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "    plot_unlabelled_version(ax, name)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# MULTI_GENOME VS N50, NUM_NODES, NUM_EDGES SCATTER\n",
    "\n",
    "ANI=99 \n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat.n50, y=plot_dat.multi_genome_node_proportion, hue=plot_dat.tool, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat.index:\n",
    "    if plot_dat.loc[i, 'tool'] == 'megahit':\n",
    "        ano = f'k={int(plot_dat.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(plot_dat.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat.n50[i], plot_dat.multi_genome_node_proportion[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_n50bymultigenome.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat.num_nodes, y=plot_dat.multi_genome_node_proportion, hue=plot_dat.tool, s=20)\n",
    "plt.xscale('log')\n",
    "texts = list()\n",
    "for i in plot_dat.index:\n",
    "    if plot_dat.loc[i, 'tool'] == 'megahit':\n",
    "        ano = f'k={int(plot_dat.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(plot_dat.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat.num_nodes[i], plot_dat.multi_genome_node_proportion[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numnodebymultigenome.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat.num_edges, y=plot_dat.multi_genome_node_proportion, hue=plot_dat.tool, s=20)\n",
    "plt.xscale('log')\n",
    "texts = list()\n",
    "for i in plot_dat.index:\n",
    "    if plot_dat.loc[i, 'tool'] == 'megahit':\n",
    "        ano = f'k={int(plot_dat.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(plot_dat.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat.num_edges[i], plot_dat.multi_genome_node_proportion[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numedgesbymultigenome.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "# MULTI_GENOME VS N50, NUM_NODES, NUM_EDGES SCATTER COPAN ONLY\n",
    "plot_dat_cpn = plot_dat.loc[plot_dat.tool == 'copangraph',:]\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat_cpn.n50, y=plot_dat_cpn.multi_genome_node_proportion, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat_cpn.index:\n",
    "    ano = f'sd={float(plot_dat_cpn.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat_cpn.n50[i], plot_dat_cpn.multi_genome_node_proportion[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_n50bymultigenome_copanonly.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat_cpn.num_nodes, y=plot_dat_cpn.multi_genome_node_proportion, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat_cpn.index:\n",
    "    ano = f'sd={float(plot_dat_cpn.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat_cpn.num_nodes[i], plot_dat_cpn.multi_genome_node_proportion[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numnodebymultigenome_copanonly.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat_cpn.num_edges, y=plot_dat_cpn.multi_genome_node_proportion, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat_cpn.index:\n",
    "    ano = f'sd={float(plot_dat_cpn.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat_cpn.num_edges[i], plot_dat_cpn.multi_genome_node_proportion[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numedgesbymultigenome_copanonly.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2518038/3516839579.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  plot_dat = data_melt.groupby(['tool', 'parameter']).apply(\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# COVERAGE VS N50, NUM_NODES, NUM_EDGES SCATTER\n",
    "ANI=99\n",
    "get_base = lambda x: os.path.splitext(os.path.basename(x))[0]\n",
    "asm_dat = pd.read_csv('ani_99_plot_dat.csv')\n",
    "data = pd.read_csv('../data/KmerMixing/gherig/ani_99_covered_positions.csv', index_col=0)\n",
    "#data.genome = [f'{g}_{i}' for i, g in zip(data.index, data.genome)]\n",
    "data['genome_ani'] = ANI\n",
    "data['tool'] = data.file.apply(lambda x: 'copangraph' if 'copangraph' in get_base(x) else 'megahit')\n",
    "data['parameter'] = data.file.apply(lambda x: float(get_base(x)[1:]) if 'copangraph' not in get_base(x) else float(get_base(x).split('_')[3]))\n",
    "data_melt = data.melt(id_vars=['tool', 'parameter', 'file', 'genome'], value_vars=['covered_bp', 'total_bp', 'proportion_covered'])\n",
    "data_melt = data_melt.merge(asm_dat, on=['tool', 'parameter', 'file'])\n",
    "## coverage by n50\n",
    "plot_dat = data_melt.groupby(['tool', 'parameter']).apply(\n",
    "    lambda x: pd.Series([\n",
    "        x.n50.iloc[0], \n",
    "        x.num_nodes.iloc[0], \n",
    "        x.num_edges.iloc[0], \n",
    "        x.loc[x.variable == 'proportion_covered'].value.mean(), \n",
    "        x.loc[x.variable == 'covered_bp'].value.sum() / x.loc[x.variable == 'total_bp'].value.sum()\n",
    "    ],\n",
    "        index = [\n",
    "        'n50',\n",
    "        'num_nodes',\n",
    "        'num_edges',\n",
    "        'macro_cov',\n",
    "        'micro_cov'\n",
    "        ]\n",
    "    )\n",
    ").reset_index()\n",
    "plot_dat\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat.n50, y=plot_dat.micro_cov, hue=plot_dat.tool, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat.index:\n",
    "    if plot_dat.loc[i, 'tool'] == 'megahit':\n",
    "        ano = f'k={int(plot_dat.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(plot_dat.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat.n50[i], plot_dat.micro_cov[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_n50bymicrocov.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "# coverage by num_nodes\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat.num_nodes, y=plot_dat.micro_cov, hue=plot_dat.tool, s=20)\n",
    "plt.xscale('log')\n",
    "texts = list()\n",
    "for i in plot_dat.index:\n",
    "    if plot_dat.loc[i, 'tool'] == 'megahit':\n",
    "        ano = f'k={int(plot_dat.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(plot_dat.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat.num_nodes[i], plot_dat.micro_cov[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numnodesbymicrocov.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name)\n",
    "\n",
    "# coverage by num_edges \n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat.num_edges, y=plot_dat.micro_cov, hue=plot_dat.tool, s=20)\n",
    "plt.xscale('log')\n",
    "texts = list()\n",
    "for i in plot_dat.index:\n",
    "    if plot_dat.loc[i, 'tool'] == 'megahit':\n",
    "        ano = f'k={int(plot_dat.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(plot_dat.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat.num_edges[i], plot_dat.micro_cov[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numedgesbymicrocov.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name)\n",
    "\n",
    "# MULTI_GENOME VS N50, NUM_NODES, NUM_EDGES SCATTER COPAN ONLY\n",
    "plot_dat_cpn = plot_dat.loc[plot_dat.tool == 'copangraph',:]\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat_cpn.n50, y=plot_dat_cpn.micro_cov, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat_cpn.index:\n",
    "    ano = f'sd={float(plot_dat_cpn.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat_cpn.n50[i], plot_dat_cpn.micro_cov[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_n50bymicrocov_copanonly.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat_cpn.num_nodes, y=plot_dat_cpn.micro_cov, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat_cpn.index:\n",
    "    ano = f'sd={float(plot_dat_cpn.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat_cpn.num_nodes[i], plot_dat_cpn.micro_cov[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numnodebymicrocov_copanonly.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=plot_dat_cpn.num_edges, y=plot_dat_cpn.micro_cov, s=20)\n",
    "texts = list()\n",
    "for i in plot_dat_cpn.index:\n",
    "    ano = f'sd={float(plot_dat_cpn.parameter[i])}'\n",
    "    plt.annotate(ano, (plot_dat_cpn.num_edges[i], plot_dat_cpn.micro_cov[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF + 'gherig_numedgesbymicrocov_copanonly.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembler</th>\n",
       "      <th>parameter</th>\n",
       "      <th>name</th>\n",
       "      <th>n50</th>\n",
       "      <th>num_nodes</th>\n",
       "      <th>num_edges</th>\n",
       "      <th>metric</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.001</td>\n",
       "      <td>gherig_copan_sd_0.001</td>\n",
       "      <td>2475</td>\n",
       "      <td>1175776</td>\n",
       "      <td>247772</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.773083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.005</td>\n",
       "      <td>gherig_copan_sd_0.005</td>\n",
       "      <td>1825</td>\n",
       "      <td>1207357</td>\n",
       "      <td>315215</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.790058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.010</td>\n",
       "      <td>gherig_copan_sd_0.01</td>\n",
       "      <td>1481</td>\n",
       "      <td>1229141</td>\n",
       "      <td>360768</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.800900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.020</td>\n",
       "      <td>gherig_copan_sd_0.02</td>\n",
       "      <td>1270</td>\n",
       "      <td>1250303</td>\n",
       "      <td>403586</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.811608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.030</td>\n",
       "      <td>gherig_copan_sd_0.03</td>\n",
       "      <td>1217</td>\n",
       "      <td>1257578</td>\n",
       "      <td>418093</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.813971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.040</td>\n",
       "      <td>gherig_copan_sd_0.04</td>\n",
       "      <td>1195</td>\n",
       "      <td>1260739</td>\n",
       "      <td>424231</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.814993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.050</td>\n",
       "      <td>gherig_copan_sd_0.05</td>\n",
       "      <td>1181</td>\n",
       "      <td>1262531</td>\n",
       "      <td>427526</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.815720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.100</td>\n",
       "      <td>gherig_copan_sd_0.1</td>\n",
       "      <td>1150</td>\n",
       "      <td>1266432</td>\n",
       "      <td>434087</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.817650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.750</td>\n",
       "      <td>gherig_copan_sd_0.75</td>\n",
       "      <td>1148</td>\n",
       "      <td>1266662</td>\n",
       "      <td>434520</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.817658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>megahit</td>\n",
       "      <td>21.000</td>\n",
       "      <td>k21.fastg</td>\n",
       "      <td>103</td>\n",
       "      <td>21383619</td>\n",
       "      <td>24414695</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.794718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>megahit</td>\n",
       "      <td>29.000</td>\n",
       "      <td>k29.fastg</td>\n",
       "      <td>295</td>\n",
       "      <td>5133689</td>\n",
       "      <td>3233694</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.702190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>megahit</td>\n",
       "      <td>39.000</td>\n",
       "      <td>k39.fastg</td>\n",
       "      <td>572</td>\n",
       "      <td>2883156</td>\n",
       "      <td>1316326</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.618752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>megahit</td>\n",
       "      <td>59.000</td>\n",
       "      <td>k59.fastg</td>\n",
       "      <td>1124</td>\n",
       "      <td>1581335</td>\n",
       "      <td>720785</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.565100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>megahit</td>\n",
       "      <td>79.000</td>\n",
       "      <td>k79.fastg</td>\n",
       "      <td>1626</td>\n",
       "      <td>1205037</td>\n",
       "      <td>604011</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.534773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>megahit</td>\n",
       "      <td>99.000</td>\n",
       "      <td>k99.fastg</td>\n",
       "      <td>2153</td>\n",
       "      <td>943184</td>\n",
       "      <td>479660</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.510939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>megahit</td>\n",
       "      <td>119.000</td>\n",
       "      <td>k119.fastg</td>\n",
       "      <td>2332</td>\n",
       "      <td>894053</td>\n",
       "      <td>423715</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.502952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>megahit</td>\n",
       "      <td>141.000</td>\n",
       "      <td>k141.fastg</td>\n",
       "      <td>2388</td>\n",
       "      <td>898483</td>\n",
       "      <td>391462</td>\n",
       "      <td>cov_F-score</td>\n",
       "      <td>0.497154</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     assembler  parameter                   name   n50  num_nodes  num_edges  \\\n",
       "0   copangraph      0.001  gherig_copan_sd_0.001  2475    1175776     247772   \n",
       "6   copangraph      0.005  gherig_copan_sd_0.005  1825    1207357     315215   \n",
       "12  copangraph      0.010   gherig_copan_sd_0.01  1481    1229141     360768   \n",
       "18  copangraph      0.020   gherig_copan_sd_0.02  1270    1250303     403586   \n",
       "24  copangraph      0.030   gherig_copan_sd_0.03  1217    1257578     418093   \n",
       "30  copangraph      0.040   gherig_copan_sd_0.04  1195    1260739     424231   \n",
       "36  copangraph      0.050   gherig_copan_sd_0.05  1181    1262531     427526   \n",
       "42  copangraph      0.100    gherig_copan_sd_0.1  1150    1266432     434087   \n",
       "48  copangraph      0.750   gherig_copan_sd_0.75  1148    1266662     434520   \n",
       "54     megahit     21.000              k21.fastg   103   21383619   24414695   \n",
       "60     megahit     29.000              k29.fastg   295    5133689    3233694   \n",
       "66     megahit     39.000              k39.fastg   572    2883156    1316326   \n",
       "72     megahit     59.000              k59.fastg  1124    1581335     720785   \n",
       "78     megahit     79.000              k79.fastg  1626    1205037     604011   \n",
       "84     megahit     99.000              k99.fastg  2153     943184     479660   \n",
       "90     megahit    119.000             k119.fastg  2332     894053     423715   \n",
       "96     megahit    141.000             k141.fastg  2388     898483     391462   \n",
       "\n",
       "         metric     value  \n",
       "0   cov_F-score  0.773083  \n",
       "6   cov_F-score  0.790058  \n",
       "12  cov_F-score  0.800900  \n",
       "18  cov_F-score  0.811608  \n",
       "24  cov_F-score  0.813971  \n",
       "30  cov_F-score  0.814993  \n",
       "36  cov_F-score  0.815720  \n",
       "42  cov_F-score  0.817650  \n",
       "48  cov_F-score  0.817658  \n",
       "54  cov_F-score  0.794718  \n",
       "60  cov_F-score  0.702190  \n",
       "66  cov_F-score  0.618752  \n",
       "72  cov_F-score  0.565100  \n",
       "78  cov_F-score  0.534773  \n",
       "84  cov_F-score  0.510939  \n",
       "90  cov_F-score  0.502952  \n",
       "96  cov_F-score  0.497154  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_records = glob.glob('../data/KmerMixing/gherig/*coverage_records.csv')\n",
    "cov_records = pd.concat(pd.read_csv(e) for e in cov_records)\n",
    "cov_records.sort_values(by=[ 'assembler', 'parameter', 'metric'], inplace=True)\n",
    "cov_records = cov_records.loc[cov_records.parameter != 0,:]\n",
    "cov_records.reset_index(drop=True, inplace=True)\n",
    "cov_records\n",
    "\n",
    "## F-score by n50\n",
    "subrecord = cov_records.loc[cov_records.metric == 'cov_F-score', :]\n",
    "#subrecord.to_csv(OUTDIR + PREF_COV + '_coverage_F-score_raw_dat.csv', index=None)\n",
    "subrecord\n",
    "#plt.figure(figsize=(4,4))\n",
    "#ax = sns.scatterplot(x=subrecord.n50, y=subrecord.value, hue=subrecord.assembler, s=20)\n",
    "#texts = list()\n",
    "#for i in subrecord.index:\n",
    "#    if subrecord.loc[i, 'assembler'] == 'megahit':\n",
    "#        ano = f'k={int(subrecord.parameter[i])}'\n",
    "#    else:\n",
    "#        ano = f'sd={float(subrecord.parameter[i])}'\n",
    "#    plt.annotate(ano, (subrecord.n50[i], subrecord.value[i]), textcoords='offset points', xytext=(2, 2), size=8)\n",
    "#name = OUTDIR + PREF_COV + 'cov_F-score_n50.pdf'\n",
    "#plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "#plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "#\n",
    "## recall by n50\n",
    "#subrecord = cov_records.loc[cov_records.metric == 'cov_recall', :]\n",
    "#subrecord\n",
    "#plt.figure(figsize=(4,4))\n",
    "#ax = sns.scatterplot(x=subrecord.n50, y=subrecord.value, hue=subrecord.assembler, s=20)\n",
    "#texts = list()\n",
    "#for i in subrecord.index:\n",
    "#    if subrecord.loc[i, 'assembler'] == 'megahit':\n",
    "#        ano = f'k={int(subrecord.parameter[i])}'\n",
    "#    else:\n",
    "#        ano = f'sd={float(subrecord.parameter[i])}'\n",
    "#    plt.annotate(ano, (subrecord.n50[i], subrecord.value[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "#name = OUTDIR + PREF_COV + 'cov_recall_n50.pdf'\n",
    "#plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "#plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "#\n",
    "## recall by n50\n",
    "#subrecord = cov_records.loc[cov_records.metric == 'cov_precsion', :]\n",
    "#subrecord\n",
    "#plt.figure(figsize=(4,4))\n",
    "#ax = sns.scatterplot(x=subrecord.n50, y=subrecord.value, hue=subrecord.assembler, s=20)\n",
    "#texts = list()\n",
    "#for i in subrecord.index:\n",
    "#    if subrecord.loc[i, 'assembler'] == 'megahit':\n",
    "#        ano = f'k={int(subrecord.parameter[i])}'\n",
    "#    else:\n",
    "#        ano = f'sd={float(subrecord.parameter[i])}'\n",
    "#    plt.annotate(ano, (subrecord.n50[i], subrecord.value[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "#name = OUTDIR + PREF_COV + 'cov_precision_n50.pdf'\n",
    "#plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "#plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "#\n",
    "## F-score by num_edges\n",
    "#subrecord = cov_records.loc[cov_records.metric == 'cov_F-score', :]\n",
    "#subrecord\n",
    "#plt.figure(figsize=(4,4))\n",
    "#ax = sns.scatterplot(x=subrecord.num_edges, y=subrecord.value, hue=subrecord.assembler, s=20)\n",
    "#ax.set_xscale('log')\n",
    "#texts = list()\n",
    "#for i in subrecord.index:\n",
    "#    if subrecord.loc[i, 'assembler'] == 'megahit':\n",
    "#        ano = f'k={int(subrecord.parameter[i])}'\n",
    "#    else:\n",
    "#        ano = f'sd={float(subrecord.parameter[i])}'\n",
    "#    plt.annotate(ano, (subrecord.num_edges[i], subrecord.value[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "#name = OUTDIR + PREF_COV + 'cov_F-score_num_edges.pdf'\n",
    "#plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "#plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "#\n",
    "## recall by num_edges \n",
    "#subrecord = cov_records.loc[cov_records.metric == 'cov_recall', :]\n",
    "#subrecord\n",
    "#plt.figure(figsize=(4,4))\n",
    "#ax = sns.scatterplot(x=subrecord.num_edges, y=subrecord.value, hue=subrecord.assembler, s=20)\n",
    "#ax.set_xscale('log')\n",
    "#texts = list()\n",
    "#for i in subrecord.index:\n",
    "#    if subrecord.loc[i, 'assembler'] == 'megahit':\n",
    "#        ano = f'k={int(subrecord.parameter[i])}'\n",
    "#    else:\n",
    "#        ano = f'sd={float(subrecord.parameter[i])}'\n",
    "#    plt.annotate(ano, (subrecord.num_edges[i], subrecord.value[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "#name = OUTDIR + PREF_COV + 'cov_recall_num_edges.pdf'\n",
    "#plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "#plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "#\n",
    "#\n",
    "## F-score by num_nodes\n",
    "#subrecord = cov_records.loc[cov_records.metric == 'cov_F-score', :]\n",
    "#subrecord\n",
    "#plt.figure(figsize=(4,4))\n",
    "#ax = sns.scatterplot(x=subrecord.num_nodes, y=subrecord.value, hue=subrecord.assembler, s=20)\n",
    "#ax.set_xscale('log')\n",
    "#texts = list()\n",
    "#for i in subrecord.index:\n",
    "#    if subrecord.loc[i, 'assembler'] == 'megahit':\n",
    "#        ano = f'k={int(subrecord.parameter[i])}'\n",
    "#    else:\n",
    "#        ano = f'sd={float(subrecord.parameter[i])}'\n",
    "#    plt.annotate(ano, (subrecord.num_nodes[i], subrecord.value[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "#name = OUTDIR + PREF_COV + 'cov_F-score_num_nodes.pdf'\n",
    "#plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "#plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "#\n",
    "## recall by num_nodes \n",
    "#subrecord = cov_records.loc[cov_records.metric == 'cov_recall', :]\n",
    "#subrecord\n",
    "#plt.figure(figsize=(4,4))\n",
    "#ax = sns.scatterplot(x=subrecord.num_nodes, y=subrecord.value, hue=subrecord.assembler, s=20)\n",
    "#ax.set_xscale('log')\n",
    "#texts = list()\n",
    "#for i in subrecord.index:\n",
    "#    if subrecord.loc[i, 'assembler'] == 'megahit':\n",
    "#        ano = f'k={int(subrecord.parameter[i])}'\n",
    "#    else:\n",
    "#        ano = f'sd={float(subrecord.parameter[i])}'\n",
    "#    plt.annotate(ano, (subrecord.num_nodes[i], subrecord.value[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "#name = OUTDIR + PREF_COV + 'cov_recall_num_nodes.pdf'\n",
    "#plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "#plot_unlabelled_version(ax, name, tight_layout=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading mix records\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembler</th>\n",
       "      <th>parameter</th>\n",
       "      <th>name</th>\n",
       "      <th>n50</th>\n",
       "      <th>num_nodes</th>\n",
       "      <th>num_edges</th>\n",
       "      <th>sum_maxlen_bp</th>\n",
       "      <th>total_ref_bps</th>\n",
       "      <th>sum_multigenome_nodes</th>\n",
       "      <th>prop_multigenome_nodes</th>\n",
       "      <th>sum_maxlen_multigenome_bp</th>\n",
       "      <th>prop_maxlen_multigenome_bp</th>\n",
       "      <th>sum_multigenome_coverage_bp</th>\n",
       "      <th>prop_multigenome_coverage_bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.001</td>\n",
       "      <td>gherig_copan_sd_0.001</td>\n",
       "      <td>2475</td>\n",
       "      <td>1175776</td>\n",
       "      <td>247772</td>\n",
       "      <td>1325970016</td>\n",
       "      <td>413817357</td>\n",
       "      <td>23283</td>\n",
       "      <td>0.019802</td>\n",
       "      <td>14729568</td>\n",
       "      <td>0.011109</td>\n",
       "      <td>25934496</td>\n",
       "      <td>0.062671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.005</td>\n",
       "      <td>gherig_copan_sd_0.005</td>\n",
       "      <td>1825</td>\n",
       "      <td>1207357</td>\n",
       "      <td>315215</td>\n",
       "      <td>1199250946</td>\n",
       "      <td>413817357</td>\n",
       "      <td>32629</td>\n",
       "      <td>0.027025</td>\n",
       "      <td>24914307</td>\n",
       "      <td>0.020775</td>\n",
       "      <td>46646006</td>\n",
       "      <td>0.112721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.010</td>\n",
       "      <td>gherig_copan_sd_0.01</td>\n",
       "      <td>1481</td>\n",
       "      <td>1229141</td>\n",
       "      <td>360768</td>\n",
       "      <td>1091140280</td>\n",
       "      <td>413817357</td>\n",
       "      <td>45709</td>\n",
       "      <td>0.037188</td>\n",
       "      <td>37046223</td>\n",
       "      <td>0.033952</td>\n",
       "      <td>74325279</td>\n",
       "      <td>0.179609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.020</td>\n",
       "      <td>gherig_copan_sd_0.02</td>\n",
       "      <td>1270</td>\n",
       "      <td>1250303</td>\n",
       "      <td>403586</td>\n",
       "      <td>999025638</td>\n",
       "      <td>413817357</td>\n",
       "      <td>64831</td>\n",
       "      <td>0.051852</td>\n",
       "      <td>51245403</td>\n",
       "      <td>0.051295</td>\n",
       "      <td>110716966</td>\n",
       "      <td>0.267550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.030</td>\n",
       "      <td>gherig_copan_sd_0.03</td>\n",
       "      <td>1217</td>\n",
       "      <td>1257578</td>\n",
       "      <td>418093</td>\n",
       "      <td>976369105</td>\n",
       "      <td>413817357</td>\n",
       "      <td>71930</td>\n",
       "      <td>0.057197</td>\n",
       "      <td>54839302</td>\n",
       "      <td>0.056167</td>\n",
       "      <td>119912112</td>\n",
       "      <td>0.289771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.040</td>\n",
       "      <td>gherig_copan_sd_0.04</td>\n",
       "      <td>1195</td>\n",
       "      <td>1260739</td>\n",
       "      <td>424231</td>\n",
       "      <td>967290372</td>\n",
       "      <td>413817357</td>\n",
       "      <td>75203</td>\n",
       "      <td>0.059650</td>\n",
       "      <td>56169640</td>\n",
       "      <td>0.058069</td>\n",
       "      <td>124155465</td>\n",
       "      <td>0.300025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.050</td>\n",
       "      <td>gherig_copan_sd_0.05</td>\n",
       "      <td>1181</td>\n",
       "      <td>1262531</td>\n",
       "      <td>427526</td>\n",
       "      <td>961796562</td>\n",
       "      <td>413817357</td>\n",
       "      <td>77312</td>\n",
       "      <td>0.061236</td>\n",
       "      <td>56835501</td>\n",
       "      <td>0.059093</td>\n",
       "      <td>126988739</td>\n",
       "      <td>0.306871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.100</td>\n",
       "      <td>gherig_copan_sd_0.1</td>\n",
       "      <td>1150</td>\n",
       "      <td>1266432</td>\n",
       "      <td>434087</td>\n",
       "      <td>952910795</td>\n",
       "      <td>413817357</td>\n",
       "      <td>81514</td>\n",
       "      <td>0.064365</td>\n",
       "      <td>57696764</td>\n",
       "      <td>0.060548</td>\n",
       "      <td>131817625</td>\n",
       "      <td>0.318541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>copangraph</td>\n",
       "      <td>0.750</td>\n",
       "      <td>gherig_copan_sd_0.75</td>\n",
       "      <td>1148</td>\n",
       "      <td>1266662</td>\n",
       "      <td>434520</td>\n",
       "      <td>952369450</td>\n",
       "      <td>413817357</td>\n",
       "      <td>81866</td>\n",
       "      <td>0.064631</td>\n",
       "      <td>57827270</td>\n",
       "      <td>0.060719</td>\n",
       "      <td>132282543</td>\n",
       "      <td>0.319664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>megahit</td>\n",
       "      <td>21.000</td>\n",
       "      <td>k21.fastg</td>\n",
       "      <td>103</td>\n",
       "      <td>21383619</td>\n",
       "      <td>24414695</td>\n",
       "      <td>1432285844</td>\n",
       "      <td>413817357</td>\n",
       "      <td>353565</td>\n",
       "      <td>0.016534</td>\n",
       "      <td>34988490</td>\n",
       "      <td>0.024428</td>\n",
       "      <td>77700839</td>\n",
       "      <td>0.187766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>megahit</td>\n",
       "      <td>29.000</td>\n",
       "      <td>k29.fastg</td>\n",
       "      <td>295</td>\n",
       "      <td>5133689</td>\n",
       "      <td>3233694</td>\n",
       "      <td>1045502506</td>\n",
       "      <td>413817357</td>\n",
       "      <td>148803</td>\n",
       "      <td>0.028986</td>\n",
       "      <td>19131616</td>\n",
       "      <td>0.018299</td>\n",
       "      <td>40513190</td>\n",
       "      <td>0.097901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>megahit</td>\n",
       "      <td>39.000</td>\n",
       "      <td>k39.fastg</td>\n",
       "      <td>572</td>\n",
       "      <td>2883156</td>\n",
       "      <td>1316326</td>\n",
       "      <td>957731834</td>\n",
       "      <td>413817357</td>\n",
       "      <td>58526</td>\n",
       "      <td>0.020299</td>\n",
       "      <td>10967219</td>\n",
       "      <td>0.011451</td>\n",
       "      <td>22986209</td>\n",
       "      <td>0.055547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>megahit</td>\n",
       "      <td>59.000</td>\n",
       "      <td>k59.fastg</td>\n",
       "      <td>1124</td>\n",
       "      <td>1581335</td>\n",
       "      <td>720785</td>\n",
       "      <td>863922259</td>\n",
       "      <td>413817357</td>\n",
       "      <td>43248</td>\n",
       "      <td>0.027349</td>\n",
       "      <td>8975100</td>\n",
       "      <td>0.010389</td>\n",
       "      <td>17990856</td>\n",
       "      <td>0.043475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>megahit</td>\n",
       "      <td>79.000</td>\n",
       "      <td>k79.fastg</td>\n",
       "      <td>1626</td>\n",
       "      <td>1205037</td>\n",
       "      <td>604011</td>\n",
       "      <td>836773838</td>\n",
       "      <td>413817357</td>\n",
       "      <td>44224</td>\n",
       "      <td>0.036699</td>\n",
       "      <td>9055815</td>\n",
       "      <td>0.010822</td>\n",
       "      <td>16840395</td>\n",
       "      <td>0.040695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>megahit</td>\n",
       "      <td>99.000</td>\n",
       "      <td>k99.fastg</td>\n",
       "      <td>2153</td>\n",
       "      <td>943184</td>\n",
       "      <td>479660</td>\n",
       "      <td>811239259</td>\n",
       "      <td>413817357</td>\n",
       "      <td>36872</td>\n",
       "      <td>0.039093</td>\n",
       "      <td>8742711</td>\n",
       "      <td>0.010777</td>\n",
       "      <td>15769751</td>\n",
       "      <td>0.038108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>megahit</td>\n",
       "      <td>119.000</td>\n",
       "      <td>k119.fastg</td>\n",
       "      <td>2332</td>\n",
       "      <td>894053</td>\n",
       "      <td>423715</td>\n",
       "      <td>827212334</td>\n",
       "      <td>413817357</td>\n",
       "      <td>33155</td>\n",
       "      <td>0.037084</td>\n",
       "      <td>8670950</td>\n",
       "      <td>0.010482</td>\n",
       "      <td>15116353</td>\n",
       "      <td>0.036529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>megahit</td>\n",
       "      <td>141.000</td>\n",
       "      <td>k141.fastg</td>\n",
       "      <td>2388</td>\n",
       "      <td>898483</td>\n",
       "      <td>391462</td>\n",
       "      <td>855559605</td>\n",
       "      <td>413817357</td>\n",
       "      <td>33085</td>\n",
       "      <td>0.036823</td>\n",
       "      <td>9247415</td>\n",
       "      <td>0.010809</td>\n",
       "      <td>15470206</td>\n",
       "      <td>0.037384</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     assembler  parameter                   name   n50  num_nodes  num_edges  \\\n",
       "0   copangraph      0.001  gherig_copan_sd_0.001  2475    1175776     247772   \n",
       "1   copangraph      0.005  gherig_copan_sd_0.005  1825    1207357     315215   \n",
       "2   copangraph      0.010   gherig_copan_sd_0.01  1481    1229141     360768   \n",
       "3   copangraph      0.020   gherig_copan_sd_0.02  1270    1250303     403586   \n",
       "4   copangraph      0.030   gherig_copan_sd_0.03  1217    1257578     418093   \n",
       "5   copangraph      0.040   gherig_copan_sd_0.04  1195    1260739     424231   \n",
       "6   copangraph      0.050   gherig_copan_sd_0.05  1181    1262531     427526   \n",
       "7   copangraph      0.100    gherig_copan_sd_0.1  1150    1266432     434087   \n",
       "8   copangraph      0.750   gherig_copan_sd_0.75  1148    1266662     434520   \n",
       "9      megahit     21.000              k21.fastg   103   21383619   24414695   \n",
       "10     megahit     29.000              k29.fastg   295    5133689    3233694   \n",
       "11     megahit     39.000              k39.fastg   572    2883156    1316326   \n",
       "12     megahit     59.000              k59.fastg  1124    1581335     720785   \n",
       "13     megahit     79.000              k79.fastg  1626    1205037     604011   \n",
       "14     megahit     99.000              k99.fastg  2153     943184     479660   \n",
       "15     megahit    119.000             k119.fastg  2332     894053     423715   \n",
       "16     megahit    141.000             k141.fastg  2388     898483     391462   \n",
       "\n",
       "    sum_maxlen_bp  total_ref_bps  sum_multigenome_nodes  \\\n",
       "0      1325970016      413817357                  23283   \n",
       "1      1199250946      413817357                  32629   \n",
       "2      1091140280      413817357                  45709   \n",
       "3       999025638      413817357                  64831   \n",
       "4       976369105      413817357                  71930   \n",
       "5       967290372      413817357                  75203   \n",
       "6       961796562      413817357                  77312   \n",
       "7       952910795      413817357                  81514   \n",
       "8       952369450      413817357                  81866   \n",
       "9      1432285844      413817357                 353565   \n",
       "10     1045502506      413817357                 148803   \n",
       "11      957731834      413817357                  58526   \n",
       "12      863922259      413817357                  43248   \n",
       "13      836773838      413817357                  44224   \n",
       "14      811239259      413817357                  36872   \n",
       "15      827212334      413817357                  33155   \n",
       "16      855559605      413817357                  33085   \n",
       "\n",
       "    prop_multigenome_nodes  sum_maxlen_multigenome_bp  \\\n",
       "0                 0.019802                   14729568   \n",
       "1                 0.027025                   24914307   \n",
       "2                 0.037188                   37046223   \n",
       "3                 0.051852                   51245403   \n",
       "4                 0.057197                   54839302   \n",
       "5                 0.059650                   56169640   \n",
       "6                 0.061236                   56835501   \n",
       "7                 0.064365                   57696764   \n",
       "8                 0.064631                   57827270   \n",
       "9                 0.016534                   34988490   \n",
       "10                0.028986                   19131616   \n",
       "11                0.020299                   10967219   \n",
       "12                0.027349                    8975100   \n",
       "13                0.036699                    9055815   \n",
       "14                0.039093                    8742711   \n",
       "15                0.037084                    8670950   \n",
       "16                0.036823                    9247415   \n",
       "\n",
       "    prop_maxlen_multigenome_bp  sum_multigenome_coverage_bp  \\\n",
       "0                     0.011109                     25934496   \n",
       "1                     0.020775                     46646006   \n",
       "2                     0.033952                     74325279   \n",
       "3                     0.051295                    110716966   \n",
       "4                     0.056167                    119912112   \n",
       "5                     0.058069                    124155465   \n",
       "6                     0.059093                    126988739   \n",
       "7                     0.060548                    131817625   \n",
       "8                     0.060719                    132282543   \n",
       "9                     0.024428                     77700839   \n",
       "10                    0.018299                     40513190   \n",
       "11                    0.011451                     22986209   \n",
       "12                    0.010389                     17990856   \n",
       "13                    0.010822                     16840395   \n",
       "14                    0.010777                     15769751   \n",
       "15                    0.010482                     15116353   \n",
       "16                    0.010809                     15470206   \n",
       "\n",
       "    prop_multigenome_coverage_bp  \n",
       "0                       0.062671  \n",
       "1                       0.112721  \n",
       "2                       0.179609  \n",
       "3                       0.267550  \n",
       "4                       0.289771  \n",
       "5                       0.300025  \n",
       "6                       0.306871  \n",
       "7                       0.318541  \n",
       "8                       0.319664  \n",
       "9                       0.187766  \n",
       "10                      0.097901  \n",
       "11                      0.055547  \n",
       "12                      0.043475  \n",
       "13                      0.040695  \n",
       "14                      0.038108  \n",
       "15                      0.036529  \n",
       "16                      0.037384  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('reading mix records')\n",
    "ms_records = glob.glob('../data/KmerMixing/gherig/*multisample_record.csv')\n",
    "ms_records = pd.concat(pd.read_csv(e) for e in ms_records)\n",
    "ms_records.rename({'param':'parameter'}, axis=1, inplace=True)\n",
    "ms_records.sort_values(by=[ 'assembler', 'parameter'], inplace=True)\n",
    "ms_records.reset_index(drop=True, inplace=True)\n",
    "\n",
    "ms_records\n",
    "\n",
    "# node-max-bp to ref\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.num_nodes, y=ms_records.prop_multigenome_coverage_bp, hue=ms_records.assembler, s=20)\n",
    "ax.set_xscale('log')\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.num_nodes[i], ms_records.prop_multigenome_coverage_bp[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'multi_genome_coverage_nodes.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.num_edges, y=ms_records.prop_multigenome_coverage_bp, hue=ms_records.assembler, s=20)\n",
    "ax.set_xscale('log')\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.num_edges[i], ms_records.prop_multigenome_coverage_bp[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'multi_genome_coverage_edges.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.n50, y=ms_records.prop_multigenome_coverage_bp, hue=ms_records.assembler, s=20)\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.n50[i], ms_records.prop_multigenome_coverage_bp[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'multi_genome_coverage_n50.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "# node-max-bp to total-in-graph\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.num_nodes, y=ms_records.prop_maxlen_multigenome_bp, hue=ms_records.assembler, s=20)\n",
    "ax.set_xscale('log')\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.num_nodes[i], ms_records.prop_maxlen_multigenome_bp[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'maxlen_multigenome_nodes.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.num_edges, y=ms_records.prop_maxlen_multigenome_bp, hue=ms_records.assembler, s=20)\n",
    "ax.set_xscale('log')\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.num_edges[i], ms_records.prop_maxlen_multigenome_bp[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'maxlen_multigenome_edges.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.n50, y=ms_records.prop_maxlen_multigenome_bp, hue=ms_records.assembler, s=20)\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.n50[i], ms_records.prop_maxlen_multigenome_bp[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'maxlen_multigenome_n50.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "# node to total-nodes\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.num_nodes, y=ms_records.prop_multigenome_nodes, hue=ms_records.assembler, s=20)\n",
    "ax.set_xscale('log')\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.num_nodes[i], ms_records.prop_multigenome_nodes[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'prop_multigenome_nodes_nodes.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.num_edges, y=ms_records.prop_multigenome_nodes, hue=ms_records.assembler, s=20)\n",
    "ax.set_xscale('log')\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.num_edges[i], ms_records.prop_multigenome_nodes[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'prop_multigenome_nodes_edges.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)\n",
    "\n",
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.scatterplot(x=ms_records.n50, y=ms_records.prop_multigenome_nodes, hue=ms_records.assembler, s=20)\n",
    "texts = list()\n",
    "for i in ms_records.index:\n",
    "    if ms_records.loc[i, 'assembler'] == 'megahit':\n",
    "        ano = f'k={int(ms_records.parameter[i])}'\n",
    "    else:\n",
    "        ano = f'sd={float(ms_records.parameter[i])}'\n",
    "    plt.annotate(ano, (ms_records.n50[i], ms_records.prop_multigenome_nodes[i]), textcoords='offset points', xytext=(5, 5), size=8)\n",
    "name = OUTDIR + PREF_MULTI + 'prop_multigenome_nodes_n50.pdf'\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name, tight_layout=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "manu-copangraph",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
