{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7ba500d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e94229a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{30026: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/30026_20M.blast.txt'],\n",
       " 30104: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/30104_20M.blast.txt'],\n",
       " 30526: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/30526_20M.blast.txt'],\n",
       " 30611: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/30611_20M.blast.txt'],\n",
       " 30918: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/30918_20M.blast.txt'],\n",
       " 30939: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/30939_20M.blast.txt'],\n",
       " 30948: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/30948_20M.blast.txt'],\n",
       " 31195: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/31195_20M.blast.txt'],\n",
       " 31280: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/31280_20M.blast.txt'],\n",
       " 31299: ['/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/31299_20M.blast.txt']}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLES = [30026, 30104, 30526, 30611, 30918, 30939, 30948, 31195, 31280, 31299]\n",
    "CAMISIMDIR = '/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/CAMISIMGraphQuality/camisim_reads/'\n",
    "TAXADIR = '/burg/pmg/users/ic2465/Projects/MANU_copangraph/data/TaxonomicResolution/'\n",
    "\n",
    "aln_files = {s: glob.glob(os.path.join(TAXADIR, f'{s}_20M.blast.txt')) for s in SAMPLES}\n",
    "aln_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a30232ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def genome_to_id_tab(s):\n",
    "    df = pd.read_csv(open(os.path.join(CAMISIMDIR, f'{s}_out/genome_to_id.tsv')), sep='\\t', header=None)\n",
    "    df.columns = ['genome_id', 'file']\n",
    "    df.file = df.file.apply(lambda x: os.path.basename(x))\n",
    "    df.index = df.file\n",
    "    return df\n",
    "\n",
    "def get_subject_table(fa, g2id):\n",
    "    file = os.path.basename(fa)\n",
    "    with open(fa) as f:\n",
    "        # get all headers\n",
    "        headers = [l for l in f if l.startswith('>')]\n",
    "    subjects = [(e.split(' ')[0][1:],file, g2id.loc[file, 'genome_id']) for e in headers]\n",
    "    return pd.DataFrame(subjects, columns=['subject', 'genome', 'genome_id'])\n",
    "\n",
    "def parse_taxonomic_profile(s):\n",
    "    prof = os.path.join(CAMISIMDIR, f'{s}_out/taxonomic_profile_0.txt')\n",
    "    with open(prof) as f:\n",
    "        lines = [e.strip() for e in f]\n",
    "        lines = lines[5:]\n",
    "        lines = [tuple(e.split('\\t')) for e in lines]\n",
    "    profile = pd.DataFrame(lines, columns=['taxid', 'taxa_rank', 'taxpath', 'taxpathsn', 'percentage', 'genome_id', 'cami_otu'])\n",
    "    profile = profile.loc[profile.taxa_rank == 'strain', :]\n",
    "    profile['species'] = profile.taxpathsn.apply(lambda x: x.split('|')[6])\n",
    "    profile['genera'] = profile.taxpathsn.apply(lambda x: x.split('|')[5])\n",
    "    profile['family'] = profile.taxpathsn.apply(lambda x: x.split('|')[4])\n",
    "    profile = profile[['taxid', 'taxa_rank', 'percentage', 'genome_id', 'species', 'genera', 'family']]\n",
    "    return profile\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3d4b52a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30026\n",
      "30104\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 8\u001b[0m\n\u001b[1;32m      6\u001b[0m g2id \u001b[38;5;241m=\u001b[39m genome_to_id_tab(s)\n\u001b[1;32m      7\u001b[0m fastas \u001b[38;5;241m=\u001b[39m [e \u001b[38;5;28;01mfor\u001b[39;00m e \u001b[38;5;129;01min\u001b[39;00m fastas \u001b[38;5;28;01mif\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mbasename(e) \u001b[38;5;129;01min\u001b[39;00m g2id\u001b[38;5;241m.\u001b[39mindex]\n\u001b[0;32m----> 8\u001b[0m subjects \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([\u001b[43mget_subject_table\u001b[49m\u001b[43m(\u001b[49m\u001b[43me\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mg2id\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m e \u001b[38;5;129;01min\u001b[39;00m fastas])\n\u001b[1;32m      9\u001b[0m subjects \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(subjects, profile, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenome_id\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     10\u001b[0m alns \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(TAXADIR, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00ms\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_20M.blast.txt\u001b[39m\u001b[38;5;124m'\u001b[39m), sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m)\n",
      "Cell \u001b[0;32mIn[14], line 12\u001b[0m, in \u001b[0;36mget_subject_table\u001b[0;34m(fa, g2id)\u001b[0m\n\u001b[1;32m      9\u001b[0m file \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mbasename(fa)\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(fa) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[1;32m     11\u001b[0m     \u001b[38;5;66;03m# get all headers\u001b[39;00m\n\u001b[0;32m---> 12\u001b[0m     headers \u001b[38;5;241m=\u001b[39m \u001b[43m[\u001b[49m\u001b[43ml\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43ml\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mf\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43ml\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstartswith\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m>\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     13\u001b[0m subjects \u001b[38;5;241m=\u001b[39m [(e\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m1\u001b[39m:],file, g2id\u001b[38;5;241m.\u001b[39mloc[file, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenome_id\u001b[39m\u001b[38;5;124m'\u001b[39m]) \u001b[38;5;28;01mfor\u001b[39;00m e \u001b[38;5;129;01min\u001b[39;00m headers]\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m pd\u001b[38;5;241m.\u001b[39mDataFrame(subjects, columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msubject\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenome\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenome_id\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "File \u001b[0;32m<frozen codecs>:322\u001b[0m, in \u001b[0;36mdecode\u001b[0;34m(self, input, final)\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "all_alns = list()\n",
    "for s in SAMPLES:\n",
    "    print(s)\n",
    "    profile = parse_taxonomic_profile(s)\n",
    "    fastas = glob.glob(os.path.join(CAMISIMDIR, f'{s}_out/genomes/GCA*.fa'))\n",
    "    g2id = genome_to_id_tab(s)\n",
    "    fastas = [e for e in fastas if os.path.basename(e) in g2id.index]\n",
    "    subjects = pd.concat([get_subject_table(e, g2id) for e in fastas])\n",
    "    subjects = pd.merge(subjects, profile, on='genome_id')\n",
    "    alns = pd.read_csv(os.path.join(TAXADIR, f'{s}_20M.blast.txt'), sep='\\t', header=None)\n",
    "    alns.columns = ['contig_name', 'subject', 'identity', 'aln_len', 'query_len']\n",
    "    alns = pd.merge(alns, subjects, on='subject')\n",
    "    alns['key'] = alns.contig_name.apply(lambda x: f'{s}:{x}')\n",
    "    alns = alns.loc[((alns.aln_len / alns.query_len) > 0.98) & ((alns.aln_len/alns.query_len) < 1.02) & (alns.identity >= 98), :]\n",
    "    alns['sample'] = s\n",
    "    all_alns.append(alns)\n",
    "\n",
    "all_alns = pd.concat(all_alns).reset_index(drop=True)\n",
    "all_alns = all_alns.drop_duplicates()\n",
    "all_alns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f7e973d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "005\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "02\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "03\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "04\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "05\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "075\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: RuntimeWarning: divide by zero encountered in log\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
      "/burg/pmg/users/ic2465/tmp/ipykernel_2739164/3283136187.py:13: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n"
     ]
    }
   ],
   "source": [
    "# parse copangraph\n",
    "sds = ['00', '001', '005', '01', '02', '03', '04', '05', '075', '10', '15', '20', '25', '30']\n",
    "alpha_divs= list()\n",
    "for sd in sds:\n",
    "    print(sd)\n",
    "    with open(os.path.join(TAXADIR, f'graph_sd{sd}.fasta')) as f:\n",
    "        data = [e.strip().split(':') for e in f if e.startswith('>')]\n",
    "        data = [(node[1:], sample[:-7], contig) for node, _, sample, _, contig, _, _, _ in data]\n",
    "\n",
    "    copan_seqs = pd.DataFrame(data, columns=['node', 'sample' ,'contig_name'])\n",
    "    copan_seqs['key'] = copan_seqs.apply(lambda x: x['sample'] + ':' + x['contig_name'], axis=1)\n",
    "    copan_alns = pd.merge(copan_seqs[['node', 'key']], all_alns, on='key')\n",
    "    alphas = copan_alns.groupby('node').apply(lambda x: pd.Series([np.log(x.species.nunique()), np.log(x.genera.nunique()), np.log(x.family.nunique())], index=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm']))\n",
    "    alphas['sd'] = sd\n",
    "    alpha_divs.append(alphas)\n",
    "alpha_divs = pd.concat(alpha_divs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5d6aff46",
   "metadata": {},
   "outputs": [],
   "source": [
    "#alpha_divs.to_csv(os.path.join(TAXADIR, 'alpha_divs.csv'))\n",
    "alpha_divs = pd.read_csv(os.path.join(TAXADIR, 'alpha_divs.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7d224eb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>subject</th>\n",
       "      <th>identity</th>\n",
       "      <th>aln_len</th>\n",
       "      <th>query_len</th>\n",
       "      <th>genome</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>taxid</th>\n",
       "      <th>taxa_rank</th>\n",
       "      <th>percentage</th>\n",
       "      <th>species</th>\n",
       "      <th>genera</th>\n",
       "      <th>family</th>\n",
       "      <th>key</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k141_1240</td>\n",
       "      <td>AP008937.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>211</td>\n",
       "      <td>211</td>\n",
       "      <td>GCA_000010145.1_ASM1014v1.fa</td>\n",
       "      <td>Genome3.0</td>\n",
       "      <td>334390.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.9545</td>\n",
       "      <td>Lactobacillus fermentum</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>30026:k141_1240</td>\n",
       "      <td>30026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k141_0</td>\n",
       "      <td>HE970764.1_2</td>\n",
       "      <td>99.745</td>\n",
       "      <td>392</td>\n",
       "      <td>392</td>\n",
       "      <td>GCA_000318035.1_ASM31803v1.fa</td>\n",
       "      <td>Genome12.0</td>\n",
       "      <td>1215914.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>Lactobacillus casei</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>30026:k141_0</td>\n",
       "      <td>30026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k141_2480</td>\n",
       "      <td>CP003799.1</td>\n",
       "      <td>99.694</td>\n",
       "      <td>327</td>\n",
       "      <td>326</td>\n",
       "      <td>GCA_000165775.3_ASM16577v3.fa</td>\n",
       "      <td>Genome4.0</td>\n",
       "      <td>880633.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.8790</td>\n",
       "      <td>Lactobacillus helveticus</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>30026:k141_2480</td>\n",
       "      <td>30026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k141_2480</td>\n",
       "      <td>CP003799.1</td>\n",
       "      <td>99.387</td>\n",
       "      <td>326</td>\n",
       "      <td>326</td>\n",
       "      <td>GCA_000165775.3_ASM16577v3.fa</td>\n",
       "      <td>Genome4.0</td>\n",
       "      <td>880633.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.8790</td>\n",
       "      <td>Lactobacillus helveticus</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>30026:k141_2480</td>\n",
       "      <td>30026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>k141_2480</td>\n",
       "      <td>CP003799.1_3</td>\n",
       "      <td>98.165</td>\n",
       "      <td>327</td>\n",
       "      <td>326</td>\n",
       "      <td>GCA_000165775.3_ASM16577v3.fa</td>\n",
       "      <td>Genome4.0</td>\n",
       "      <td>880633.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.8790</td>\n",
       "      <td>Lactobacillus helveticus</td>\n",
       "      <td>Lactobacillus</td>\n",
       "      <td>Lactobacillaceae</td>\n",
       "      <td>30026:k141_2480</td>\n",
       "      <td>30026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090743</th>\n",
       "      <td>k141_10856</td>\n",
       "      <td>CP001810.1_2</td>\n",
       "      <td>99.520</td>\n",
       "      <td>417</td>\n",
       "      <td>417</td>\n",
       "      <td>GCA_000145035.1_ASM14503v1.fa</td>\n",
       "      <td>Genome30.0</td>\n",
       "      <td>515622.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.0898</td>\n",
       "      <td>Butyrivibrio proteoclasticus</td>\n",
       "      <td>Butyrivibrio</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>31299:k141_10856</td>\n",
       "      <td>31299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090751</th>\n",
       "      <td>k141_10857</td>\n",
       "      <td>CP000728.1</td>\n",
       "      <td>99.978</td>\n",
       "      <td>4499</td>\n",
       "      <td>4499</td>\n",
       "      <td>GCA_000017065.1_ASM1706v1.fa</td>\n",
       "      <td>Genome18.0</td>\n",
       "      <td>441772.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.5587</td>\n",
       "      <td>Clostridium botulinum</td>\n",
       "      <td>Clostridium</td>\n",
       "      <td>Clostridiaceae</td>\n",
       "      <td>31299:k141_10857</td>\n",
       "      <td>31299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090755</th>\n",
       "      <td>k141_10858</td>\n",
       "      <td>CP018623.1</td>\n",
       "      <td>99.688</td>\n",
       "      <td>321</td>\n",
       "      <td>321</td>\n",
       "      <td>GCA_001908255.1_ASM190825v1.fa</td>\n",
       "      <td>Genome24.0</td>\n",
       "      <td>1311.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.2073</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>31299:k141_10858</td>\n",
       "      <td>31299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090756</th>\n",
       "      <td>k141_10859</td>\n",
       "      <td>CP003939.1</td>\n",
       "      <td>99.961</td>\n",
       "      <td>2551</td>\n",
       "      <td>2551</td>\n",
       "      <td>GCA_001010825.1_ASM101082v1.fa</td>\n",
       "      <td>Genome20.0</td>\n",
       "      <td>1224745.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.4154</td>\n",
       "      <td>Clostridioides difficile</td>\n",
       "      <td>Clostridioides</td>\n",
       "      <td>Peptostreptococcaceae</td>\n",
       "      <td>31299:k141_10859</td>\n",
       "      <td>31299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090759</th>\n",
       "      <td>k141_10860</td>\n",
       "      <td>CP014228.1</td>\n",
       "      <td>99.757</td>\n",
       "      <td>411</td>\n",
       "      <td>411</td>\n",
       "      <td>GCA_001553565.1_ASM155356v1.fa</td>\n",
       "      <td>Genome34.0</td>\n",
       "      <td>111015.1</td>\n",
       "      <td>strain</td>\n",
       "      <td>0.0285</td>\n",
       "      <td>Actinomyces radicidentis</td>\n",
       "      <td>Actinomyces</td>\n",
       "      <td>Actinomycetaceae</td>\n",
       "      <td>31299:k141_10860</td>\n",
       "      <td>31299</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>273746 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        contig_name       subject  identity  aln_len  query_len  \\\n",
       "0         k141_1240    AP008937.1   100.000      211        211   \n",
       "2            k141_0  HE970764.1_2    99.745      392        392   \n",
       "3         k141_2480    CP003799.1    99.694      327        326   \n",
       "4         k141_2480    CP003799.1    99.387      326        326   \n",
       "7         k141_2480  CP003799.1_3    98.165      327        326   \n",
       "...             ...           ...       ...      ...        ...   \n",
       "1090743  k141_10856  CP001810.1_2    99.520      417        417   \n",
       "1090751  k141_10857    CP000728.1    99.978     4499       4499   \n",
       "1090755  k141_10858    CP018623.1    99.688      321        321   \n",
       "1090756  k141_10859    CP003939.1    99.961     2551       2551   \n",
       "1090759  k141_10860    CP014228.1    99.757      411        411   \n",
       "\n",
       "                                 genome   genome_id      taxid taxa_rank  \\\n",
       "0          GCA_000010145.1_ASM1014v1.fa   Genome3.0   334390.1    strain   \n",
       "2         GCA_000318035.1_ASM31803v1.fa  Genome12.0  1215914.1    strain   \n",
       "3         GCA_000165775.3_ASM16577v3.fa   Genome4.0   880633.1    strain   \n",
       "4         GCA_000165775.3_ASM16577v3.fa   Genome4.0   880633.1    strain   \n",
       "7         GCA_000165775.3_ASM16577v3.fa   Genome4.0   880633.1    strain   \n",
       "...                                 ...         ...        ...       ...   \n",
       "1090743   GCA_000145035.1_ASM14503v1.fa  Genome30.0   515622.1    strain   \n",
       "1090751    GCA_000017065.1_ASM1706v1.fa  Genome18.0   441772.1    strain   \n",
       "1090755  GCA_001908255.1_ASM190825v1.fa  Genome24.0     1311.1    strain   \n",
       "1090756  GCA_001010825.1_ASM101082v1.fa  Genome20.0  1224745.1    strain   \n",
       "1090759  GCA_001553565.1_ASM155356v1.fa  Genome34.0   111015.1    strain   \n",
       "\n",
       "         percentage                       species          genera  \\\n",
       "0            0.9545       Lactobacillus fermentum   Lactobacillus   \n",
       "2            0.0543           Lactobacillus casei   Lactobacillus   \n",
       "3            0.8790      Lactobacillus helveticus   Lactobacillus   \n",
       "4            0.8790      Lactobacillus helveticus   Lactobacillus   \n",
       "7            0.8790      Lactobacillus helveticus   Lactobacillus   \n",
       "...             ...                           ...             ...   \n",
       "1090743      0.0898  Butyrivibrio proteoclasticus    Butyrivibrio   \n",
       "1090751      0.5587         Clostridium botulinum     Clostridium   \n",
       "1090755      0.2073      Streptococcus agalactiae   Streptococcus   \n",
       "1090756      0.4154      Clostridioides difficile  Clostridioides   \n",
       "1090759      0.0285      Actinomyces radicidentis     Actinomyces   \n",
       "\n",
       "                        family               key  sample  \n",
       "0             Lactobacillaceae   30026:k141_1240   30026  \n",
       "2             Lactobacillaceae      30026:k141_0   30026  \n",
       "3             Lactobacillaceae   30026:k141_2480   30026  \n",
       "4             Lactobacillaceae   30026:k141_2480   30026  \n",
       "7             Lactobacillaceae   30026:k141_2480   30026  \n",
       "...                        ...               ...     ...  \n",
       "1090743        Lachnospiraceae  31299:k141_10856   31299  \n",
       "1090751         Clostridiaceae  31299:k141_10857   31299  \n",
       "1090755       Streptococcaceae  31299:k141_10858   31299  \n",
       "1090756  Peptostreptococcaceae  31299:k141_10859   31299  \n",
       "1090759       Actinomycetaceae  31299:k141_10860   31299  \n",
       "\n",
       "[273746 rows x 15 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#all_alns.to_csv(os.path.join(TAXADIR, 'all_alns.csv'))\n",
    "all_alns = pd.read_csv(os.path.join(TAXADIR, 'all_alns.csv'), index_col=0)\n",
    "all_alns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c20bfd50",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dat = pd.melt(alpha_divs.reset_index(), value_vars=['alpha_div_sp', 'alpha_div_gn', 'alpha_div_fm'], id_vars=['node', 'sd'])\n",
    "plot_dat.sd = plot_dat.sd.apply(lambda x: float(f'0.{x}'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "805616f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.04372653084981931"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot_dat.loc[(plot_dat.sd == 0.1) & (plot_dat.variable == 'alpha_div_fm'),:]['value'].replace(-np.inf, np.nan).dropna().std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0555c4eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_unlabelled_version(ax, name, tight_layout=True):\n",
    "    name = os.path.splitext(name)[0]\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    frame1 = plt.gca()\n",
    "    frame1.legend().set_visible(False)\n",
    "    if tight_layout:\n",
    "        plt.tight_layout()\n",
    "    plt.savefig(f'{name}_UNLBLD.pdf', dpi=1400, bbox_inches='tight')\n",
    "    plt.savefig(f'{name}_UNLBLD.png', dpi=900, bbox_inches='tight')\n",
    "    plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "04c2c468",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 400x400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(4,4))\n",
    "ax = sns.lineplot(x=plot_dat.sd, y=plot_dat.value, hue=plot_dat.variable)\n",
    "plt.yscale('log')\n",
    "name = os.path.join(TAXADIR, 'taxonomy_curve.pdf')\n",
    "ax.set_ylabel('alpha-diversity')\n",
    "plt.savefig(name, dpi=1400, bbox_inches='tight')\n",
    "plot_unlabelled_version(ax, name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df78472d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    222914.000000\n",
       "mean          0.004425\n",
       "std           0.059882\n",
       "min           0.000000\n",
       "25%           0.000000\n",
       "50%           0.000000\n",
       "75%           0.000000\n",
       "max           2.302585\n",
       "dtype: float64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kspecies_alpha.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "4768438c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.log10(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "manu-copangraph",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
